Implement JSON-RPC server library

feuGeneA opened the pull request:
Fixes https://github.com/NomicFoundation/rethnet/issues/92

Wodann commented on crates/rethnet_rpc_server/src/lib.rs:
```diff
@@ -1 +1,200 @@
+
+type StateType = Arc<Mutex<HybridState<RethnetLayer>>>;
+
+pub async fn router(state: StateType) -> Router {
```
I recommend following the pattern used in the readme of axum: https://crates.io/crates/axum/

You can add types to the function signature that tell it what type of Json object you're expecting, reducing boilerplate code

![image](https://github.com/NomicFoundation/hardhat/assets/6917585/9839f971-21cb-4dde-a963-074b338b15fb)

feuGeneA replied:
i could use some more input on what exactly you'd like to see happen here. my understanding is that we'll only have a single "handler", because we'll only ever route to one "endpoint": `/`. so, i'm not sure what benefit there would be to breaking it out into a separate function like this. i did make a small change to have `axum` do the JSON parsing instead of doing it ourselves. see that [here](https://github.com/NomicFoundation/hardhat/pull/3998/commits/afc1a6d0be2cc29bf3303d6ba27568371053856b). is that what you had in mind?

Wodann replied:
Yes, we'll have a single handler, but by using the code as described in the example, it automatically handles and responds with an error for the failure case when parsing. It reduces boilerplate for you.

feuGeneA replied:
i see, but i'm still not seeing how you'd like to see it applied to the code here, especially now that i've introduced all the `handle_*` functions. if you think there's still room for further improvement beyond the introduction of those functions, let's discuss on a call.

Wodann commented on crates/rethnet_rpc_server/src/lib.rs:
```diff
@@ -1 +1,200 @@
+    use revm::primitives::state::AccountInfo;
+    use std::net::{SocketAddr, TcpListener};
+
+    fn start_server() -> SocketAddr {
```
I'd move part of this function into the lib as `async fn run(addr: SocketAddr, genesis_accounts: HashMap<Address, AccountInfo>)`, which runs the `axum::Server`. We'll need that once we wrap this in the `rethnet` CLI too.

feuGeneA replied:
addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/07c267afd1162739fbe13cc16ce74af6c5e56238

Wodann commented on crates/rethnet_rpc_server/src/lib.rs:
```diff
@@ -1 +1,200 @@
+        .expect("should decode response from UTF-8")
+    }
+
+    #[tokio::test]
```
Given that these are integration tests - using the public API of rethnet_rpc_server - you can use the integration test folder: `crates/rethnet_rpc_server/tests`

See https://doc.rust-lang.org/cargo/guide/project-layout.html

feuGeneA replied:
but, _are_ these integration tests? i'm happy to defer to your judgement, but these actually felt like unit tests to me, and that's why i kept them here. i know that sometimes people say things like, "if it's doing network interaction then it's an integration test." but to me, an integration test is a test that exercises the integration of multiple different components. and to me that seems to be the reason for having the integration test folder in the rust project hierarchy: it provides a place to put tests that combine multiple different components, because such tests would be out of place within the `mod tests` sub-module of a single component, because they're not specific to any one component. here, however, all of the tests are specific to the contents of this `lib` module, so i felt they belonged within that module. again, i'm happy to defer to your preference, but that was my rationale.

Wodann replied:
I generally agree with what you said, but based on your description I'd recognise four components at work here:

1. Serialization
2. Client (reqwest)
3. Server (listener + handler)
4. Deserialization

As such, I'd say that it qualifies as an integration test under your definition too.

The Rust doc says something along similar lines: https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

feuGeneA replied:
addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/37b288e819d30d5913a8cb7425bc47657ec59912

Wodann commented on crates/rethnet_rpc_server/src/lib.rs:
```diff
@@ -1 +1,200 @@
+
+    async fn submit_request(address: &SocketAddr, request: &RpcRequest) -> String {
+        String::from_utf8(
+            hyper::body::to_bytes(
```
Why not use `reqwest`, which we already have as a dependency?

feuGeneA replied:
i actually forgot about that. i just modeled this after the `axum` examples. but, if i understand correctly, this isn't introducing an additional dependency, because `hyper` is already a dependency of `axum`. do you think there's a good reason to pick one over the other for these tests?

Wodann replied:
You do actually add the `hyper` dependency in the cargo toml. If not for this test, would you need it there?

feuGeneA replied:
you're right, thanks for pushing. addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/77ee710ca3d23139066c1ede9c62e0fe0091ccc8

Wodann commented on crates/rethnet_rpc_server/Cargo.toml:
```diff
@@ -4,10 +4,16 @@ version = "0.1.0"
+hyper = "0.14.26"
 rethnet_eth = { version = "0.1.0-dev", path = "../rethnet_eth", features = ["serde"] }
+rethnet_evm = { version = "0.1.0-dev", path = "../rethnet_evm" }
+rethnet_evm_napi = { version = "0.1.0-dev", path = "../rethnet_evm_napi" }
```
We should not need napi here

feuGeneA replied:
https://github.com/NomicFoundation/hardhat/pull/3998/commits/394d688b2f2f5e705615ac46e6f5eb6a16147b48

feuGeneA replied:
in an early iterations i had tried out a direct instantiation of the runtime object in rethnet_evm_napi, but i quickly realized that wasn't right, but this old dependency was left unnoticed. thanks for pointing it out.

Wodann commented on crates/rethnet_rpc_server/Cargo.toml:
```diff
@@ -4,10 +4,16 @@ version = "0.1.0"
 rethnet_eth = { version = "0.1.0-dev", path = "../rethnet_eth", features = ["serde"] }
+rethnet_evm = { version = "0.1.0-dev", path = "../rethnet_evm" }
+rethnet_evm_napi = { version = "0.1.0-dev", path = "../rethnet_evm_napi" }
+revm = { git = "https://github.com/Wodann/revm", rev = "a4550cc", version = "3.1", default-features = false, features = ["dev", "secp256k1", "serde", "std"] }
```
What do we need `revm` for?

feuGeneA replied:
i was using it in a few places but it was unnecessary and i eliminated those references (and this dependency) in https://github.com/NomicFoundation/hardhat/pull/3998/commits/6c036a9c5c614a8f227d2a8ecaebc10bc73cab78

feuGeneA commented:
@Wodann the overlap of changes between this PR and #3998 was too much to manage them independently, so i just pushed all of the changes from here over to there. i addressed some of your comments here with commits in that other PR, and i'll link to them from here for your convenience. some of your comments here haven't been addressed with changes yet, but i'm going to continue the conversation here where you started them. let's add any new comments to #3998 instead of here.

feuGeneA commented on crates/rethnet_rpc_server/src/lib.rs:
```diff
@@ -1 +1,200 @@
+
 mod methods;
+
+type StateType = Arc<Mutex<HybridState<RethnetLayer>>>;
```
addressed in https://github.com/NomicFoundation/hardhat/pull/3998/commits/d39c5c522deb6c1f5c725ff0d6e4cf5d2f437a1b
